### PR TITLE
Share equivalent RoPE cache storage

### PR DESCRIPTION
**Pull request by Mystri**

### CODE REVIEW

File: torchtitan/models/common/rope.py [hunk 1]
```diff
@@ -18,9 +20,107 @@
...
+    def __init__(self, registry: "_RoPECacheRegistry", slot_name: str) -> None:
+        self._registry = registry
+        self._slot_name = slot_name
+
+    def read(self) -> torch.Tensor:
+        return self._registry._read(self._slot_name)
+
+
+@contextlib.contextmanager
+def _rope_cache_registry_context(registry: "_RoPECacheRegistry"):
```
Comment (tianyu-l) [inline torchtitan/models/common/rope.py]: why do we need the capability of switch "registry", instead of always only having one default

File: torchtitan/models/common/rope.py [hunk 2]
```diff
@@ -110,7 +210,51 @@ class Config(Module.Config):
     def __init__(self, config: Config):
         super().__init__()
         self.config = config
-        self.register_buffer("cache", self._precompute_cache(), persistent=False)
+        self._cache_reader: RoPECacheReader | None = None
+        cache = self._precompute_cache()
```
Comment (tianyu-l) [inline torchtitan/models/common/rope.py]: although memory is saved, compute seems still wasted

File: torchtitan/models/common/rope.py
Comment (tianyu-l) [inline torchtitan/models/common/rope.py:1]: Does it cover https://github.com/pytorch/torchtitan/blob/main/torchtitan/overrides/helion_rope.py -- it looks so because cache construction code is shared.

File: torchtitan/models/deepseek_v3/mtp.py [hunk 3]
```diff
@@ -212,13 +212,14 @@ def __init__(self, config: Config):
...
 
         self.mtp_layers = ModuleList()
-        for layer_config in config.mtp_layers:
-            if not isinstance(layer_config, MTPTransformerBlock.Config):
-                raise ValueError(
-                    "MTPDecoder requires Config.mtp_layers to contain "
-                    "MTPTransformerBlock.Config instances."
-                )
-            self.mtp_layers.append(layer_config.build())
+        with self._rope_cache_context():
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/mtp.py]: related to the other comment -- why MTP requires an explicit context

**Comment by tianyu-l** [comment]
@claude 
Previously each attention module would instantiate its own RoPE cache, which would be a problem for memory for long context training. This is the second attempt, previous one being https://github.com/pytorch/torchtitan/pull/4111

Personally I like this one better since its cache management is explicit.

Please review this PR. In particular
- check if my comments make sense
- compare the two solutions and give your preference

**Comment by tianyu-l** [comment]
@claude 

Can't we just keep the cache at the RoPE module level, without touching Decoder at all?

**Comment by tianyu-l** [comment]
@claude 
I see. I think this issue exists only because we are doing
- first meta-device init
- and then parallelization

If we reverse it, we could avoid calling `.to()` or `to_empty()`. Is there caveat around doing that?

Another doubt:
Letting decoder owns the cache sounds less straightforward than explicitly passing rope cache to each attention module, instead of letting each attention owns its own rope module.

**Comment by tianyu-l** [comment]
@claude 

> And passing a shared rope module (rather than the cache) is worse, for the reasons in the earlier review: the same module appears under multiple attention FQNs, breaking "one module owns its state," duplicating checkpoint FQNs, and confusing FSDP/PP traversal. Sharing the buffer while modules stay independent is the right layer to dedup at.

I don't believe there's a lot of heterogenous rope used in the same model. It's usually at most one. In that case, how about
- make `rope` an Decoder-owned module, with cache still as registered buffer
- explicitly pass it into the constructors of `Attention` modules
- change PP stage building so that the `Decoder.rope` exists on every PP stage (multiple PP stages on the same PP rank owns duplicate and we admit this overhead)

**Comment by tianyu-l** [comment]
@Mystri see if you agree with https://github.com/pytorch/torchtitan/pull/4376#issuecomment-5465560593

**Comment by Mystri** [comment]
> @Mystri see if you agree with [#4376 (comment)](https://github.com/pytorch/torchtitan/pull/4376#issuecomment-5465560593)

We cannot safely assert homogeneity for newer models such as deepseek-v4, which computes `freqs_cis` with different rope_theta for compressed/non-compressed layers:
https://huggingface.co/deepseek-ai/DeepSeek-V4-Pro/blob/main/inference/model.py#L473

My proposal that fits yours best is that the `Decoder` creates and owns both types of ropes and pass them as a "dict of ropes" along, and the builder extracts the correct rope module from it when building the submodule.
@claude

File: torchtitan/models/common/rope.py
Comment (Mystri) [reply torchtitan/models/common/rope.py:1]: It should, rope module construction is now managed separately from using it - register it into Decoder and use its reference on each Attention layer

**Comment by Mystri** [comment]
> @Mystri see if you agree with [#4376 (comment)](https://github.com/pytorch/torchtitan/pull/4376#issuecomment-5465560593)

@tianyu-l just updated code design according your comments, please check it out.

File: torchtitan/models/common/rope.py
Comment (tianyu-l) [inline torchtitan/models/common/rope.py:190]: it's done here

File: torchtitan/models/common/rope.py [hunk 4]
```diff
@@ -167,13 +244,13 @@ def forward(
...
     def _init_self_buffers(self, *, buffer_device: torch.device | None = None) -> None:
-        # TODO: In long-term we need to have buffer abstraction in `Module`` class to infer the buffer_device
         if buffer_device is None:
             # After ``to_empty()``, the existing cache records the target device.
             # Recompute there when the caller does not pass an explicit buffer device.
             buffer_device = self.cache.device
         with torch.device(buffer_device):
-            self.cache = self._precompute_cache()
+            cache = self._precompute_cache()
+            self.register_buffer("cache", cache, persistent=False)
```
Comment (tianyu-l) [inline torchtitan/models/common/rope.py:253]: do we need to do it here again?

File: torchtitan/models/common/decoder.py [hunk 5]
```diff
@@ -253,12 +248,27 @@ def update_from_config(
...
 
+        self.rope_modules = ModuleDict()
+        for layer_config in config.layers:
+            attention_config = getattr(layer_config, "attention", None)
+            rope_config = getattr(attention_config, "rope", None)
+            if rope_config is None:
+                continue
+            key = rope_config.rope_key()
+            if key not in self.rope_modules:
+                self.rope_modules[key] = rope_config.build()
```
Comment (tianyu-l) [inline torchtitan/models/common/decoder.py:261]: I found this fallback to be unnecessary

File: torchtitan/models/deepseek_v3/model.py [hunk 6]
```diff
@@ -92,7 +90,8 @@ def __init__(self, config: Config):
             self.softmax_scale = self.softmax_scale * mscale * mscale
 
         self.inner_attention = config.inner_attention.build()
-        self.rope = config.rope.build()
+        # Keep the canonical module registered only under Decoder.rope_modules.
+        object.__setattr__(self, "rope", rope_modules[config.rope.rope_key()])
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/model.py:94]: how does this compare with
```
self.rope = rope_modules[config.rope.rope_key()]
```

File: torchtitan/models/deepseek_v3/model.py [hunk 7]
```diff
@@ -164,9 +163,9 @@ class DeepSeekV3TransformerBlock(TransformerBlock):
     class Config(TransformerBlock.Config):
         pass
 
-    def __init__(self, config: Config):
+    def __init__(self, config: Config, *, rope_modules: ModuleDict):
         super().__init__()
-        self.attention = config.attention.build()
+        self.attention = config.attention.build(rope_modules=rope_modules)
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/model.py:168]: actually, what's the main difference between sending it to constructor, or passing it through attention `forward`?

File: torchtitan/models/common/attention.py
Comment (tianyu-l) [inline torchtitan/models/common/attention.py:907]: I don't think we need this / should keep a complete rope config.

It sounds pretty similar to attention
- For (most) common models, there's only one attention_mask needed
- For some models, there'd be more than one
- We can make `attention_masks` a union of things, optionally just one mask
- DSv4 can determine which one to use based on if there's "compress ratio"

E.g. here's what we do for full attention vs. sliding window attention for gpt-oss https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/gpt_oss/model.py#L146

File: torchtitan/models/common/attention.py [hunk 8]
```diff
@@ -917,7 +921,7 @@ def __post_init__(self) -> None:
                     f"n_kv_heads ({n_kv_heads})"
                 )
 
-    def __init__(self, config: Config):
+    def __init__(self, config: Config, *, rope_modules: ModuleDict):
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py:924]: To mimic gpt-oss attention, here we only need to send the single relevant rope module to attention, from transformer block. In other words, rope module selection is done at transformer block, not attention. (but doesn't feel that different)

File: torchtitan/models/common/rope.py [hunk 9]
```diff
@@ -107,6 +110,80 @@ class Config(Module.Config):
         original_seq_len: int = 4096
         truncate: bool = True
 
+        def rope_key(self) -> str:
```
Comment (tianyu-l) [inline torchtitan/models/common/rope.py:113]: Are you trying to build a robust cache key? I think most model would have limited number of rope cache, so this feels over-complicated, compared with a simple string based key.

**Comment by tianyu-l** [comment]
@claude review latest changes and my comments

File: docs/rope-cache-updated-plan.md
Comment (tianyu-l) [inline docs/rope-cache-updated-plan.md:1]: please remove this

**Comment by tianyu-l** [comment]
would you like to follow up?
